frontend/src/hooks/useAsyncData.ts

```diff
@@ -1,31 +1,135 @@
 /* Copyright 2024 Marimo. All rights reserved. */
-import type { SetStateAction } from "jotai";
-import { type DependencyList, useState, useEffect, type Dispatch } from "react";
+import {
+  type DependencyList,
+  type Dispatch,
+  type SetStateAction,
+  useState,
+  useEffect,
+} from "react";
 import useEvent from "react-use-event-hook";
+import { invariant } from "@/utils/invariant";
 
-interface AsyncDataResponse<T> {
-  data: T | undefined;
-  loading: boolean;
-  error: Error | undefined;
-  setData: Dispatch<SetStateAction<T | undefined>>;
-  reload: () => void;
+interface PendingResult {
+  status: "pending";
+  data: undefined;
+  error: undefined;
+  isPending: true;
+  isFetching: true;
 }
 
-// eslint-disable-next-line @typescript-eslint/no-explicit-any
-export function combineAsyncData<T extends any[]>(
-  ...responses: { [K in keyof T]: AsyncDataResponse<T[K]> }
-): Omit<AsyncDataResponse<T>, "setData"> {
-  return {
-    data: responses.every((response) => response.data !== undefined)
-      ? (responses.map((response) => response.data) as T)
-      : undefined,
-    loading: responses.some((response) => response.loading),
-    error:
-      responses.find((response) => response.error !== null)?.error || undefined,
-    reload: () => {
-      responses.forEach((response) => response.reload());
-    },
+interface LoadingResult<T> {
+  status: "loading";
+  data: T;
+  error: undefined;
+  isPending: false;
+  isFetching: true;
+}
+
+interface ErrorResult {
+  status: "error";
+  data: undefined;
```

Comment:
the last regression (and happy to be convinced that it is ok) is that if you get an error, you still may want to show the previous `data`. for example in a table, you may want to show the old data but put the error in the banner (showing both)

Reply:
Oh yeah, you're right. To my best estimation, we don't rely on this right now, but React Query allows for errors to have the `data` filled with `placeholderData` (carried over from some previous state). I updated our `ErrorResult` to carry over a previous data (if any). It allows for this kind of pattern:


```ts
if (isPending) {
   return <Loader />
}

if (error) {
  return <ErrorComponent error={error} maybeData={data} />
}

return <Component data={data} />
```